File: python_src/Implementations/my_graph.py
#Graph implementation

class Graph():
	def __init__(self):
		self.nodes = 0
		self.adj_list = {}
		
	def addVertex(self, node):
		self.nodes += 1
		self.adj_list[node] = []
		return self.adj_list
			
	def addEdge(self, node1, node2):
		self.adj_list[node1].append(node2)
		self.adj_list[node2].append(node1)
		return self.adj_list
	
	def showConnections(self):
		all_nodes = self.adj_list.keys()
		for node in all_nodes:
			node_connections = self.adj_list[node]
			connections = ''
			for vertex in node_connections:
				connections += vertex + ' '
			print(f'{node} --> {connections}')
			
myGraph = Graph()
myGraph.addVertex("0")
myGraph.addVertex("1")
myGraph.addVertex("2")
myGraph.addVertex("3")
myGraph.addVertex("4")
myGraph.addVertex("5")
myGraph.addVertex("6")
myGraph.addEdge("3", "1")
myGraph.addEdge("3", "4")
myGraph.addEdge("4", "2")
myGraph.addEdge("4", "5")
myGraph.addEdge("1", "2")
myGraph.addEdge("1", "0")
myGraph.addEdge("0", "2")
myGraph.addEdge("6", "5")

myGraph.showConnections()

File: cdsa/my_graph.c
#include "my_graph.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

static int _find(Graph *self, const char *label) {
    for (int i = 0; i < self->count; i++)
        if (strcmp(self->vertices[i].label, label) == 0) return i;
    return -1;
}

static void _adj_append(Vertex *v, const char *label) {
    AdjNode *node = (AdjNode *)malloc(sizeof(AdjNode));
    strncpy(node->label, label, MAX_LABEL - 1);
    node->label[MAX_LABEL - 1] = '\0';
    node->next = NULL;
    if (!v->head) { v->head = node; return; }
    AdjNode *cur = v->head;
    while (cur->next) cur = cur->next;
    cur->next = node;
}

static void _add_vertex(Graph *self, const char *label) {
    if (self->count >= MAX_VERTICES) return;
    strncpy(self->vertices[self->count].label, label, MAX_LABEL - 1);
    self->vertices[self->count].label[MAX_LABEL - 1] = '\0';
    self->vertices[self->count].head = NULL;
    self->count++;
}

static void _add_edge(Graph *self, const char *l1, const char *l2) {
    int i1 = _find(self, l1), i2 = _find(self, l2);
    if (i1 < 0 || i2 < 0) return;
    _adj_append(&self->vertices[i1], l2);
    _adj_append(&self->vertices[i2], l1);
}

static void _show_connections(Graph *self) {
    for (int i = 0; i < self->count; i++) {
        printf("%s --> ", self->vertices[i].label);
        AdjNode *cur = self->vertices[i].head;
        while (cur) { printf("%s ", cur->label); cur = cur->next; }
        printf("\n");
    }
}

static void _destroy(Graph *self) {
    for (int i = 0; i < self->count; i++) {
        AdjNode *cur = self->vertices[i].head;
        while (cur) { AdjNode *next = cur->next; free(cur); cur = next; }
    }
    free(self);
}

Graph *Graph_new() {
    Graph *g            = (Graph *)malloc(sizeof(Graph));
    g->count            = 0;
    g->add_vertex       = _add_vertex;
    g->add_edge         = _add_edge;
    g->show_connections = _show_connections;
    g->destroy          = _destroy;
    return g;
}
